syntax = 'proto3';

package pbrt.protobuf;

message Point3f {
  float x = 1;
  float y = 2;
  float z = 3;
}

message Matrix {
  repeated float m = 1;
}

message AnimatedTransform {
  Matrix start_transform = 1;
  Matrix end_transform = 2;
  float start_time = 3;
  float end_time = 4;
}

message Bounds3f {
  Point3f point_min = 1;
  Point3f point_max = 2;
}

message TransformedPrimitive {
  AnimatedTransform transform = 1;
  int64 root_ref = 2;
}

message BVHNode {
  Bounds3f bounds = 1;
  int64 left_ref = 2;
  int64 right_ref = 3;
  uint32 axis = 4;

  repeated TransformedPrimitive transformed_primitives = 7;
}
